src/index.ts

```diff
@@ -0,0 +1,267 @@
+ * returns property of the query element.
+ *
+ * @param name Property name
+ * @return Property value
+ */
+export function prop<N extends keyof Element>(name: string): MatcherFn<Element[N]>;
+
+/**
+ * Generates a function which matches node of type selector, returning an
+ * attribute by property if the attribute exists. If no selector is passed,
+ * returns property of the query element.
+ *
+ * @param selector Optional selector
+ * @param name Property name
+ * @return Property value
+ */
+export function prop<N extends keyof Element>(
+	selector: string | undefined,
+	name: N
+): MatcherFn<Element[N]>;
+
+/**
+ * Generates a function which matches node of type selector, returning an
+ * attribute by property if the attribute exists. If no selector is passed,
+ * returns property of the query element.
+ *
+ * @param selector Optional selector
+ * @param name Property name
+ * @return Property value
+ */
+export function prop<N extends keyof Element>(
+	arg1: string | undefined,
+	arg2?: string
+): MatcherFn<Element[N]> {
+	let name: string;
+	let selector: string | undefined;
+	if (1 === arguments.length) {
+		name = arg1 as string;
+		selector = undefined;
+	} else {
+		name = arg2 as string;
+		selector = arg1;
+	}
+	return function (node: Element): Element[N] | undefined {
+		let match: Element | null = node;
+		if (selector) {
+			match = node.querySelector(selector);
+		}
+		if (match) {
+			return getPath(match, name) as Element[N] | undefined;
+		}
+	} as MatcherFn<Element[N]>;
+}
+
+/**
+ * Generates a function which matches node of type selector, returning an
+ * attribute by name if the attribute exists. If no selector is passed,
+ * returns attribute of the query element.
+ *
+ * @param name Attribute name
+ * @return Attribute value
+ */
+export function attr(name: string): MatcherFn<string>;
+
+/**
+ * Generates a function which matches node of type selector, returning an
+ * attribute by name if the attribute exists. If no selector is passed,
+ * returns attribute of the query element.
+ *
+ * @param selector Optional selector
+ * @param name Attribute name
+ * @return Attribute value
+ */
+export function attr(selector: string | undefined, name: string): MatcherFn<string>;
+
+/**
+ * Generates a function which matches node of type selector, returning an
+ * attribute by name if the attribute exists. If no selector is passed,
+ * returns attribute of the query element.
+ *
+ * @param selector Optional selector
+ * @param name Attribute name
+ * @return Attribute value
+ */
+export function attr(arg1: string | undefined, arg2?: string): MatcherFn<string> {
+	let name: string;
+	let selector: string | undefined;
+	if (1 === arguments.length) {
+		name = arg1 as string;
+		selector = undefined;
+	} else {
+		name = arg2 as string;
+		selector = arg1;
+	}
+	return function (node: Element): string | undefined {
+		const attributes = prop(selector, 'attributes')(node);
```

Comment:
This right here is amazing @aduth, it is actually typed as a `NamedNodeMap|undefined` without any type assertions! I had a big smile when I deleted the assertion and the type held, what a good feeling! 🎉